### PR TITLE
[Kernel] Multi-LoRA MoE CUDA BGMV kernel

### PR DESCRIPTION
**Pull request by taehokim20**
Co-authored-by: Claude

## Purpose
This PR adds custom CUDA BGMV kernels for MoE LoRA (shrink + expand), replacing the Triton SGMV kernels used in vllm. The CUDA kernels are optimized for decode-heavy MoE LoRA workloads with the following techniques:
* Multi-pair decode batching (PPB=4): Processes 4 token-expert pairs per thread block during decode, amortizing weight tile loading overhead. Prefill uses PPB=1 since the grid already saturates the GPU.
* Deep async pipeline (3 stages on sm_80+): Uses cuda::pipeline with 3-stage async memory copies for decode, overlapping global memory loads with computation. Prefill uses 2 stages to reduce shared memory pressure.
* RANK_TILE tiling: Each X input tile is loaded once into shared memory and reused across 8 weight rows (RANK_TILE=8), reducing global memory loads for X by 8×.
* Shared memory utilization: Decode path uses 216 KB of the 228 KB available on H100/H200 (95% utilization).
* Vectorized loads (up to vec_size=8): Memory accesses use up to 8-element vectorized loads (16 bytes for bf16), with fallback to smaller vec_size when alignment requires it.
* Zero-copy weight access: The kernel works directly with vLLM's native LoRA weight layout [max_loras, num_experts, rank, feat] using a lora_stride parameter, avoiding transposed weight copies that would otherwise duplicate all LoRA weights on GPU.

The kernels support all LoRA rank values (8, 16, 32, 64) and cover model dimensions for GPT-OSS-120B and Nemotron-Nano-3-30B-A3B including TP-sharded variants.

## Test Plan
* End-to-end benchmarks with 8 LoRA adapters (max_lora_rank=32), sonnet dataset, concurrency sweep [1, 2, 4, 8, 16] (targeting small batch size), measured on H100 and H200.

## Test Result
End-to-end OTPS p50 (higher is better):
**Nemotron-Nano-3-30B-A3B**
image

**GPT-OSS-120B**
image

---

### CODE REVIEW

File: vllm/lora/punica_wrapper/punica_gpu.py [hunk 1]
```diff
@@ -411,35 +768,49 @@ def add_lora_fused_moe(
...
+          - VLLM_MOE_LORA_DECODE_BACKEND  (default: cuda)
+          - VLLM_MOE_LORA_DECODE_THRESHOLD (default: 0 = always)
         """
+        # Skip entirely during cudagraph capture -- neither CUDA nor
+        # Triton MoE LoRA ops are capture-safe (they allocate GPU
+        # memory or use data-dependent tensor sizes).  During capture
+        # the model runs with dummy data and no real LoRA adapters,
+        # so the LoRA contribution is zero.
+        if torch.cuda.is_current_stream_capturing():
+            return
```
[the remark this answers is not in the document]
Comment (taehokim20) [reply vllm/lora/punica_wrapper/punica_gpu.py]: This is intentional and matches vLLM's existing LoRA + CUDA graph design. vLLM applies LoRA contributions outside the captured graph, not inside it. The Triton MoE LoRA path has the same behavior - LoRA ops are never part of the captured graph. See `cudagraph_specialize_lora` in vLLM's config and the `CudagraphDispatcher` for how LoRA is handled during graph replay. Making these kernels capture-safe is a separate effort that would apply to both the Triton and CUDA path equally.

File: csrc/lora/moe_lora_ops.cu [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+  switch (pack_u32(feat_in, feat_out)) {
+#define CASE_MOE_SHRINK(in_T, out_T, W_T, narrow, wide)                    \
+  case pack_u32(wide, narrow):                                              \
+    moe_bgmv_shrink_sliced<wide, narrow, in_T, out_T, W_T>(                \
+        Y, X, w_ptr, sorted_token_ids, expert_ids, lora_indices,            \
+        num_pairs, num_slices, num_experts, num_tokens, lora_stride, 1.0f); \
+    return true;
+    FOR_MOE_ALL_WIDE_NARROW(CASE_MOE_SHRINK, T, T, T)
+#undef CASE_MOE_SHRINK
```
[the remark this answers is not in the document]
Comment (taehokim20) [reply csrc/lora/moe_lora_ops.cu]: The scale=1.0 is correct. vLLM pre-scales LoRA weights by `alpha/rank` during weight loading (see `LoRALayerWeights`), so the kernel doesn't need a runtime scale parameter. This matches the convention used by the existing Triton MoE LoRA kernels and all other LoRA kernels in vLLM.

File: vllm/lora/punica_wrapper/punica_gpu.py [hunk 3]
```diff
@@ -72,6 +117,65 @@ def __init__(
...
+        self._moe_shrink_out_bf16 = torch.zeros(
+            max_slices, max_pairs, max_rank,
+            dtype=torch.bfloat16, device=device)
+        self._moe_shrink_out_fp16 = torch.zeros(
+            max_slices, max_pairs, max_rank,
+            dtype=torch.float16, device=device)
+        self._moe_y_accum = torch.zeros(
+            max_num_batched_tokens * max_feat_out,
+            dtype=torch.float32, device=device)
+        self._moe_y_accum_max_feat = max_feat_out
```
[the remark this answers is not in the document]
Comment (taehokim20) [reply vllm/lora/punica_wrapper/punica_gpu.py]: The 28672 dimensions in `moe_bgmv_config.h` are the `feat_in` for the shrink kernel (model hidden dim), while `max_feat_out` here covers the LoRA expand output dimension, which is <= 16384 for all currently supported models (GPT-OSS-120B, Nemotron-Super, Qwen3-30B, Nemotron-Nano). So this is safe for now, but I agree bumping the constant to 28672 would be good for future-proofing.

File: vllm/lora/punica_wrapper/punica_gpu.py [hunk 4]
```diff
@@ -390,6 +507,246 @@ def moe_lora_align_block_size(
...
+            y_orig.add_(y_accum)
+
+    def _use_cuda_for_moe_lora(self, num_tokens: int) -> bool:
+        """Decide whether to use CUDA BGMV or Triton for this call."""
+        if _MOE_DECODE_THRESHOLD > 0:
+            if num_tokens <= _MOE_DECODE_THRESHOLD:
+                return _MOE_DECODE_USE_CUDA
+            else:
+                return _MOE_PREFILL_USE_CUDA
+        return _MOE_PREFILL_USE_CUDA
```
[the remark this answers is not in the document]
Comment (taehokim20) [reply vllm/lora/punica_wrapper/punica_gpu.py]: When the threshold is 0 (default), both `VLLM_MOE_LORA_PREFILL_BACKEND` and `VLLM_MOE_LORA_DECODE_BACKEND` default to `cuda`, so returning `_MOE_PREFILL_USE_CUDA` produces the correct result -- CUDA is used for both phases. The separate env vars are provided as opt-in overrides for debugging or A/B comparison. The threshold adds an optional batch-size cutoff on top of that. The default path is intentionally simple: use CUDA for everything.

**Review by jeejeelee — CHANGES_REQUESTED** [review]
Sorry for the delayed response. I noticed that you submitted a similar PR for FI — I also think integrating this kernel into FI is better.
